File: Heap/Python/Smallest Range Covering Elements from K Lists.py
# Problem link: https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description
# Time: O(n log k)
# Space: O(k)
class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        minHeap, maxNum, small, large, size = [], -1, 0, 0, float('inf')

        for idx, arr in enumerate(nums):
            minHeap.append((arr[0], 0, idx))
            maxNum = max(maxNum, arr[0])

        heapq.heapify(minHeap)

        while len(minHeap) >= len(nums):
            if size > maxNum - minHeap[0][0]:
                size = maxNum - minHeap[0][0]
                small, large = minHeap[0][0], maxNum

            value, ptr, idx = heapq.heappop(minHeap)

            if ptr + 1 < len(nums[idx]):
                newVal = nums[idx][ptr + 1]
                maxNum = max(maxNum, newVal)
                heapq.heappush(minHeap, (newVal, ptr + 1, idx))
        
        return [small, large]

File: Heap/Java/Smallest Range Covering Elements from K Lists.java
// Problem link: https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description
// Time: O(n log k)
// Space: O(k)
class Solution {
    public int[] smallestRange(List<List<Integer>> nums) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        int maxNum = -1, small = 0, large = 0, size = Integer.MAX_VALUE, length = nums.size();

        for(int idx = 0; idx < length; idx++) {
            minHeap.add(new int[] {nums.get(idx).get(0), 0, idx});
            maxNum = Math.max(maxNum, nums.get(idx).get(0));
        }

        while(minHeap.size() >= length) {
            if(size > maxNum - minHeap.peek()[0]) {
                size = maxNum - minHeap.peek()[0];
                small = minHeap.peek()[0];
                large = maxNum;
            }

            int[] popped = minHeap.poll();

            if(popped[1] + 1 < nums.get(popped[2]).size()) {
                int newVal = nums.get(popped[2]).get(popped[1] + 1);
                minHeap.add(new int[] {newVal, popped[1] + 1, popped[2]});
                maxNum = Math.max(maxNum, newVal);
            }
        }

        return new int[] {small, large};
    } 
}
